### PR TITLE
Export data quality metric

### PR DESCRIPTION
**Pull request by james-s-willis**
I think this is ready to review, though I'm not sure on the config file change.

### CODE REVIEW

File: lib/stages/DataQuality.cpp [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+DataQuality::~DataQuality() {}
+
+void DataQuality::calc_alpha_coeffs(dset_id_t ds_id) {
+
+    auto& dm = datasetManager::instance();
+
+    auto sstate_fut = std::async(&datasetManager::dataset_state<stackState>, &dm, ds_id);
+    const stackState* ss = sstate_fut.get();
```
Comment (jrs65) [inline lib/stages/DataQuality.cpp:51]: I think there's no point doing an async call here as you don't do anything in the middle. Might as well just do a normal blocking call.

File: lib/stages/DataQuality.cpp [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        }
+
+        // Get correct set of alpha coefficients
+        const std::vector<double>& alpha = dset_id_map.at(ds_id);
+
+        // Compute sensitivity
+        double sensitivity = 0;
+
+        for (uint32_t i = 0; i < frame.num_prod; i++) {
+            auto var = (frame.weight[i] == 0 ? 0.0 : frame.weight[i]);
```
Comment (jrs65) [inline lib/stages/DataQuality.cpp:108]: I'm pretty sure this should be:
```suggestion
            auto var = (frame.weight[i] == 0 ? 0.0 : 1.0 / frame.weight[i]);
```

File: lib/stages/DataQuality.cpp [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+void DataQuality::calc_alpha_coeffs(dset_id_t ds_id) {
+
+    auto& dm = datasetManager::instance();
+
+    auto sstate_fut = std::async(&datasetManager::dataset_state<stackState>, &dm, ds_id);
+    const stackState* ss = sstate_fut.get();
+
+    if (ss == nullptr) {
+        FATAL_ERROR("Couldn't find stackState ancestor of dataset "
```
Comment (jrs65) [inline lib/stages/DataQuality.cpp:54]: You should use the `fingerprint` technique in here. The way you have this implemented at the moment, you're going  to regenerate the alphas every time the dataset ID changes, which is every 10s at the moment, but stack state never changes.

In theory you could just use the stack state hash for that purpose, but we may eventually recalculate when the flags change too, so I would go straight to the fingerprint method.

File: lib/stages/DataQuality.hpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+ *
+ *
+ * @par Buffers
+ * @buffer in_buf Visibility data.
+ *     @buffer_format VisBuffer structured
+ *     @buffer_metadata VisMetadata
+ *
+ * @conf  num_elements          Int. The number of elements (i.e. inputs) in the
+ *                              correlator data.
+ *
```
Comment (jrs65) [inline lib/stages/DataQuality.hpp:32]: Make sure to document the metrics exported.

**Review by jrs65 — CHANGES_REQUESTED** [review]

**Comment by james-s-willis** [comment]
@jrs65 I've addressed your review comments.

**Review by jrs65 — CHANGES_REQUESTED** [review]
Oh oops. I write this review yesterday, but forgot to send it!

File: lib/stages/DataQuality.cpp [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+REGISTER_KOTEKAN_STAGE(DataQuality);
+
+DataQuality::DataQuality(Config& config_, const std::string& unique_name,
+                         bufferContainer& buffer_container) :
+    Stage(config_, unique_name, buffer_container, std::bind(&DataQuality::main_thread, this)),
+    data_quality_metric(Metrics::instance().add_gauge("kotekan_dataquality_dataquality",
+                                                      unique_name, {"freq_id"})) {
+
+    // Apply config.
+    _num_elements = config.get<uint32_t>(unique_name, "num_elements");
```
[the remark this answers is not in the document]
Comment (jrs65) [reply lib/stages/DataQuality.cpp:39]: Ah okay. I think there was a mistake in my description of the algorithm. I'll fix it in the parent issue, but basically: alpha = counts**2 / sum(counts)**2

[the remark this answers is not in the document]
Comment (jrs65) [reply lib/stages/DataQuality.cpp:39]: So I don't think you need `_num_elements` because of that.

**Comment by james-s-willis** [comment]
@jrs65 I think I have fixed it now

File: lib/stages/DataQuality.cpp [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    while (!stop_thread) {
+
+        // Get input visibilities. We assume the shape of these doesn't change.
+        if (wait_for_full_frame(in_buf, unique_name.c_str(), input_frame_id) == nullptr) {
+            break;
+        }
+
+        auto frame = VisFrameView(in_buf, input_frame_id);
+        dset_id_t ds_id = frame.dataset_id;
+        auto fprint = dm.fingerprint(ds_id, {"stack"});
```
Comment (jrs65) [inline lib/stages/DataQuality.cpp:96]: I would consider populating a ds_id -> fprint map here, though I don't think it's a must do. The fingerprint calculation involves some json manipulation which probably isn't great to do every single time. It's probably fine for our purposes at the moment, though if you decide not to change it we should look at the performance while testing.

**Review by jrs65 — COMMENTED** [review]
Looking good. The only thing left is to change to the correct algorithm for the alpha as my initial version had a mistake (sorry!), and then remove `_num_elements`. Details are in a comment.

**Comment by james-s-willis** [comment]
@jrs65 I've added a map of `dataset_id` -> `fingerprint` and fixed the alpha calculation.

**Review by jrs65 — APPROVED** [review]
Great work. Thanks @james-s-willis

**Review by andrerenard — APPROVED** [review]

**Pull request merged**